0.8.19 exposes Guava in a Groovy API but has it as an implementation dependency

Hello @ejona86 and team!

Using Gradle version 7.4.1.

We have a custom Gradle plugin written in Groovy that uses this plugin as a dependency (essentially it wraps this plugin with some additional tasks).

Small sample of the code:
```
import com.google.protobuf.gradle.GenerateProtoTask

@CompileStatic
class ProtobufPlugin implements Plugin<Project> {
// plugin body
}
```

Up until `0.8.19`, we were able to annotate the `Plugin` implementation class with `@CompileStatic` as shown above. As of `0.8.19`, this is no longer working, with the following exception:

```
> Task :compileGroovy FAILED
startup failed:
General error during canonicalization: java.lang.NoClassDefFoundError: com.google.common.collect.ImmutableList

java.lang.RuntimeException: java.lang.NoClassDefFoundError: com.google.common.collect.ImmutableList
        at org.codehaus.groovy.control.CompilationUnit$IPrimaryClassNodeOperation.doPhaseOperation(CompilationUnit.java:976)
        at org.codehaus.groovy.control.CompilationUnit.processPhaseOperations(CompilationUnit.java:671)
        at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:635)
        at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:610)
        at org.gradle.api.internal.tasks.compile.ApiGroovyCompiler.execute(ApiGroovyCompiler.java:270)
        at org.gradle.api.internal.tasks.compile.ApiGroovyCompiler.execute(ApiGroovyCompiler.java:64)
        at org.gradle.api.internal.tasks.compile.GroovyCompilerFactory$DaemonSideCompiler.execute(GroovyCompilerFactory.java:97)
        at org.gradle.api.internal.tasks.compile.GroovyCompilerFactory$DaemonSideCompiler.execute(GroovyCompilerFactory.java:76)
        at org.gradle.api.internal.tasks.compile.daemon.AbstractDaemonCompiler$CompilerWorkAction.execute(AbstractDaemonCompiler.java:135)
        at org.gradle.workers.internal.DefaultWorkerServer.execute(DefaultWorkerServer.java:63)
        at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:49)
        at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:43)
        at org.gradle.internal.classloader.ClassLoaderUtils.executeInClassloader(ClassLoaderUtils.java:97)
        at org.gradle.workers.internal.AbstractClassLoaderWorker.executeInClassLoader(AbstractClassLoaderWorker.java:43)
        at org.gradle.workers.internal.IsolatedClassloaderWorker.run(IsolatedClassloaderWorker.java:49)
        at org.gradle.workers.internal.IsolatedClassloaderWorker.run(IsolatedClassloaderWorker.java:30)
        at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:87)
        at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:56)
        at org.gradle.process.internal.worker.request.WorkerAction$1.call(WorkerAction.java:138)
        at org.gradle.process.internal.worker.child.WorkerLogEventListener.withWorkerLoggingProtocol(WorkerLogEventListener.java:41)
        at org.gradle.process.internal.worker.request.WorkerAction.run(WorkerAction.java:135)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
        at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
        at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
        at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NoClassDefFoundError: com.google.common.collect.ImmutableList
        at org.codehaus.groovy.ast.decompiled.AsmReferenceResolver.resolveClass(AsmReferenceResolver.java:46)
        at org.codehaus.groovy.ast.decompiled.AsmReferenceResolver.resolveNonArrayType(AsmReferenceResolver.java:79)
        at org.codehaus.groovy.ast.decompiled.AsmReferenceResolver.resolveType(AsmReferenceResolver.java:70)
        at org.codehaus.groovy.ast.decompiled.MemberSignatureParser.createMethodNode(MemberSignatureParser.java:50)
        at org.codehaus.groovy.ast.decompiled.DecompiledClassNode.lambda$createMethodNode$1(DecompiledClassNode.java:230)
        at org.codehaus.groovy.ast.decompiled.DecompiledClassNode.createMethodNode(DecompiledClassNode.java:236)
        at org.codehaus.groovy.ast.decompiled.DecompiledClassNode.lazyInitMembers(DecompiledClassNode.java:203)
        at org.codehaus.groovy.ast.decompiled.DecompiledClassNode.getDeclaredMethods(DecompiledClassNode.java:122)
        at org.codehaus.groovy.ast.ClassNode.getDeclaredMethods(ClassNode.java:850)
        at org.codehaus.groovy.ast.ClassNode.getMethods(ClassNode.java:865)
        at org.codehaus.groovy.transform.stc.StaticTypeCheckingVisitor.findMethodsWithGenerated(StaticTypeCheckingVisitor.java:4650)
        at org.codehaus.groovy.transform.stc.StaticTypeCheckingVisitor.findMethod(StaticTypeCheckingVisitor.java:4741)
        at org.codehaus.groovy.transform.stc.StaticTypeCheckingVisitor.visitMethodCallExpression(StaticTypeCheckingVisitor.java:3439)
        at org.codehaus.groovy.transform.sc.StaticCompilationVisitor.visitMethodCallExpression(StaticCompilationVisitor.java:414)
        at org.codehaus.groovy.ast.expr.MethodCallExpression.visit(MethodCallExpression.java:76)
        at org.codehaus.groovy.ast.CodeVisitorSupport.visitExpressionStatement(CodeVisitorSupport.java:117)
        at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitExpressionStatement(ClassCodeVisitorSupport.java:200)
        at org.codehaus.groovy.transform.stc.StaticTypeCheckingVisitor.visitExpressionStatement(StaticTypeCheckingVisitor.java:2148)
        at org.codehaus.groovy.ast.stmt.ExpressionStatement.visit(ExpressionStatement.java:40)
        at org.codehaus.groovy.ast.CodeVisitorSupport.visitBlockStatement(CodeVisitorSupport.java:86)
        at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitBlockStatement(ClassCodeVisitorSupport.java:164)
        at org.codehaus.groovy.transform.stc.StaticTypeCheckingVisitor.visitBlockStatement(StaticTypeCheckingVisitor.java:3943)
        at org.codehaus.groovy.ast.stmt.BlockStatement.visit(BlockStatement.java:69)
        at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitClassCodeContainer(ClassCodeVisitorSupport.java:138)
        at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitConstructorOrMethod(ClassCodeVisitorSupport.java:111)
        at org.codehaus.groovy.transform.stc.StaticTypeCheckingVisitor.visitConstructorOrMethod(StaticTypeCheckingVisitor.java:2137)
        at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitMethod(ClassCodeVisitorSupport.java:106)
        at org.codehaus.groovy.transform.stc.StaticTypeCheckingVisitor.startMethodInference(StaticTypeCheckingVisitor.java:2567)
        at org.codehaus.groovy.transform.stc.StaticTypeCheckingVisitor.visitMethod(StaticTypeCheckingVisitor.java:2530)
        at org.codehaus.groovy.transform.sc.StaticCompilationVisitor.visitMethod(StaticCompilationVisitor.java:239)
        at org.codehaus.groovy.ast.ClassNode.visitMethods(ClassNode.java:1099)
        at org.codehaus.groovy.ast.ClassNode.visitContents(ClassNode.java:1092)
        at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitClass(ClassCodeVisitorSupport.java:52)
        at org.codehaus.groovy.transform.stc.StaticTypeCheckingVisitor.visitClass(StaticTypeCheckingVisitor.java:410)
        at org.codehaus.groovy.transform.sc.StaticCompilationVisitor.visitClass(StaticCompilationVisitor.java:197)
        at org.codehaus.groovy.transform.sc.StaticCompileTransformation.visit(StaticCompileTransformation.java:67)
        at org.codehaus.groovy.transform.ASTTransformationVisitor.visitClass(ASTTransformationVisitor.java:143)
        at org.codehaus.groovy.transform.ASTTransformationVisitor.lambda$addPhaseOperations$2(ASTTransformationVisitor.java:220)
        at org.codehaus.groovy.control.CompilationUnit$IPrimaryClassNodeOperation.doPhaseOperation(CompilationUnit.java:942)
        ... 34 more
```

It looks like some issue related to guava's `ImmutableList`, but I don't understand why it's impacting the compilation of our plugin since `ImmutableList` is not in this plugin's public API (only exposed as private or package-private).

Removing `@CompileStatic` resolves the issue, however we'd rather not lose the static compilation performance benefit.

--- a/build.gradle
+++ b/build.gradle
@@ -56,7 +56,6 @@ configurations {
 dependencies {
   compileOnly "com.android.tools.build:gradle:4.1.0"
 
-  implementation 'com.google.guava:guava:27.0.1-jre'
   implementation 'com.google.gradle:osdetector-gradle-plugin:1.7.0'
 
   testImplementation 'junit:junit:4.12'

--- a/src/main/groovy/com/google/protobuf/gradle/ExecutableLocator.groovy
+++ b/src/main/groovy/com/google/protobuf/gradle/ExecutableLocator.groovy
@@ -28,7 +28,6 @@
  */
 package com.google.protobuf.gradle
 
-import com.google.common.base.Preconditions
 import groovy.transform.CompileStatic
 import groovy.transform.PackageScope
 import org.gradle.api.Named

--- a/src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy
+++ b/src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy
@@ -31,8 +31,6 @@ package com.google.protobuf.gradle
 
 import static java.nio.charset.StandardCharsets.US_ASCII
 
-import com.google.common.base.Preconditions
-import com.google.common.collect.ImmutableList
 import groovy.transform.CompileStatic
 import groovy.transform.PackageScope
 import groovy.transform.TypeChecked
@@ -103,7 +101,7 @@ public abstract class GenerateProtoTask extends DefaultTask {
   transient private SourceSet sourceSet
   @SuppressWarnings("UnnecessaryTransientModifier") // It is not necessary for task to implement Serializable
   transient private Object variant
-  private ImmutableList<String> flavors
+  private List<String> flavors
   private String buildType
   private boolean isTestVariant
   private final Provider<Boolean> isAndroidProject = providerFactory.provider { Utils.isAndroidProject(project) }
@@ -275,11 +273,11 @@ public abstract class GenerateProtoTask extends DefaultTask {
     this.isTestVariant = isTestVariant
   }
 
-  void setFlavors(ImmutableList<String> flavors) {
+  void setFlavors(List<String> flavors) {
     checkInitializing()
     Preconditions.checkState(isAndroidProject.get(),
         'flavors should not be set in a Java project')
-    this.flavors = flavors
+    this.flavors = Collections.unmodifiableList(new ArrayList<String>(flavors))
   }
 
   void setBuildType(String buildType) {
@@ -370,7 +368,7 @@ public abstract class GenerateProtoTask extends DefaultTask {
   }
 
   @Internal("Not an actual input to the task, only used to find tasks belonging to a variant")
-  ImmutableList<String> getFlavors() {
+  List<String> getFlavors() {
     Preconditions.checkState(isAndroidProject.get(),
         'flavors should not be used in a Java project')
     Preconditions.checkNotNull(flavors, 'flavors is not set')

--- a/src/main/groovy/com/google/protobuf/gradle/Preconditions.java
+++ b/src/main/groovy/com/google/protobuf/gradle/Preconditions.java
@@ -1,0 +1,55 @@
+/*
+ * Copyright (c) 2022, Google Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors
+ * may be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.protobuf.gradle;
+
+/**
+ * Common assertions helper.
+ */
+final class Preconditions {
+  private Preconditions() {} // prevent instantiation
+
+  public static void checkState(boolean expectedState) {
+    if (!expectedState) {
+      throw new IllegalStateException();
+    }
+  }
+
+  public static void checkState(boolean expectedState, Object errorMessage) {
+    if (!expectedState) {
+      throw new IllegalStateException(String.valueOf(errorMessage));
+    }
+  }
+
+  public static <T> T checkNotNull(T obj, Object errorMessage) {
+    if (obj == null) {
+      throw new NullPointerException(String.valueOf(errorMessage));
+    }
+    return obj;
+  }
+}

--- a/src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy
+++ b/src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy
@@ -29,7 +29,6 @@
  */
 package com.google.protobuf.gradle
 
-import com.google.common.collect.ImmutableList
 import groovy.transform.CompileStatic
 import groovy.transform.TypeChecked
 import groovy.transform.TypeCheckingMode
@@ -320,7 +319,7 @@ class ProtobufPlugin implements Plugin<Project> {
       Provider<GenerateProtoTask> generateProtoTask = addGenerateProtoTask(
           variant.name, sourceDirs, extractProtosDirs, extractIncludeProtosTask) {
         it.setVariant(variant, isTestVariant)
-        it.flavors = ImmutableList.copyOf(variant.productFlavors.collect { it.name } )
+        it.flavors = variant.productFlavors.collect { it.name }
         if (variant.hasProperty('buildType')) {
           it.buildType = variant.buildType.name
         }

--- a/src/main/groovy/com/google/protobuf/gradle/Utils.groovy
+++ b/src/main/groovy/com/google/protobuf/gradle/Utils.groovy
@@ -28,7 +28,6 @@
  */
 package com.google.protobuf.gradle
 
-import com.google.common.base.Preconditions
 import groovy.transform.CompileStatic
 import org.gradle.api.Project
 import org.gradle.api.tasks.SourceSet

--- a/src/test/groovy/com/google/protobuf/gradle/IDESupportTest.groovy
+++ b/src/test/groovy/com/google/protobuf/gradle/IDESupportTest.groovy
@@ -28,7 +28,6 @@
  */
 package com.google.protobuf.gradle
 
-import com.google.common.collect.ImmutableSet
 import groovy.transform.CompileDynamic
 import org.gradle.testkit.runner.BuildResult
 import org.gradle.testkit.runner.GradleRunner
@@ -80,25 +79,25 @@ class IDESupportTest extends Specification {
       }
     }
 
-    Set<String> expectedSourceDir = ImmutableSet.builder()
-      .add('file://$MODULE_DIR$/src/main/java')
-      .add('file://$MODULE_DIR$/src/grpc/proto')
-      .add('file://$MODULE_DIR$/src/main/proto')
-      .add('file://$MODULE_DIR$/build/extracted-include-protos/grpc')
-      .add('file://$MODULE_DIR$/build/extracted-protos/main')
-      .add('file://$MODULE_DIR$/build/extracted-include-protos/main')
-      .add('file://$MODULE_DIR$/build/extracted-protos/grpc')
-      .add('file://$MODULE_DIR$/build/generated/source/proto/grpc/java')
-      .add('file://$MODULE_DIR$/build/generated/source/proto/grpc/grpc_output')
-      .add('file://$MODULE_DIR$/build/generated/source/proto/main/java')
-      .build()
-    Set<String> expectedTestSourceDir = ImmutableSet.builder()
-      .add('file://$MODULE_DIR$/src/test/java')
-      .add('file://$MODULE_DIR$/src/test/proto')
-      .add('file://$MODULE_DIR$/build/extracted-protos/test')
-      .add('file://$MODULE_DIR$/build/extracted-include-protos/test')
-      .add('file://$MODULE_DIR$/build/generated/source/proto/test/java')
-      .build()
+    Set<String> expectedSourceDir = [
+      'file://$MODULE_DIR$/src/main/java',
+      'file://$MODULE_DIR$/src/grpc/proto',
+      'file://$MODULE_DIR$/src/main/proto',
+      'file://$MODULE_DIR$/build/extracted-include-protos/grpc',
+      'file://$MODULE_DIR$/build/extracted-protos/main',
+      'file://$MODULE_DIR$/build/extracted-include-protos/main',
+      'file://$MODULE_DIR$/build/extracted-protos/grpc',
+      'file://$MODULE_DIR$/build/generated/source/proto/grpc/java',
+      'file://$MODULE_DIR$/build/generated/source/proto/grpc/grpc_output',
+      'file://$MODULE_DIR$/build/generated/source/proto/main/java',
+    ]
+    Set<String> expectedTestSourceDir = [
+      'file://$MODULE_DIR$/src/test/java',
+      'file://$MODULE_DIR$/src/test/proto',
+      'file://$MODULE_DIR$/build/extracted-protos/test',
+      'file://$MODULE_DIR$/build/extracted-include-protos/test',
+      'file://$MODULE_DIR$/build/generated/source/proto/test/java',
+    ]
     Set<String> expectedGeneratedDirs = [
       'file://$MODULE_DIR$/build/extracted-include-protos/grpc',
       'file://$MODULE_DIR$/build/extracted-protos/main',
@@ -154,14 +153,14 @@ class IDESupportTest extends Specification {
       }
     }
 
-    Set<String> expectedSourceDir = ImmutableSet.builder()
-      .add('src/main/java')
-      .add('src/test/java')
-      .add('build/generated/source/proto/grpc/java')
-      .add('build/generated/source/proto/grpc/grpc_output')
-      .add('build/generated/source/proto/main/java')
-      .add('build/generated/source/proto/test/java')
-      .build()
+    Set<String> expectedSourceDir = [
+      'src/main/java',
+      'src/test/java',
+      'build/generated/source/proto/grpc/java',
+      'build/generated/source/proto/grpc/grpc_output',
+      'build/generated/source/proto/main/java',
+      'build/generated/source/proto/test/java',
+    ]
     assert Objects.equals(expectedSourceDir, sourceDir)
 
     where:

